pkg/controllers/proxy/hairpin_controller.go

```diff
@@ -0,0 +1,121 @@
+package proxy
+
+import (
+	"fmt"
+	"net"
+	"os"
+	"path"
+	"runtime"
+	"sync"
+	"time"
+
+	"github.com/cloudnativelabs/kube-router/v2/pkg/healthcheck"
+	"github.com/cloudnativelabs/kube-router/v2/pkg/utils"
+	"github.com/vishvananda/netns"
+	"k8s.io/klog/v2"
+)
+
+type hairpinController struct {
+	epC <-chan string
+	nsc *NetworkServicesController
+}
+
+func (hpc *hairpinController) Run(stopCh <-chan struct{}, wg *sync.WaitGroup,
+	healthChan chan<- *healthcheck.ControllerHeartbeat) {
+	defer wg.Done()
+	klog.Infof("Starting hairping controller (handles setting hairpin_mode for veth interfaces)")
+
+	t := time.NewTicker(healthcheck.HPCSyncPeriod)
+	defer t.Stop()
+	for {
+		// Add an additional non-blocking select to ensure that if the stopCh channel is closed it is handled first
+		select {
+		case <-stopCh:
+			klog.Info("Shutting down Hairpin Controller goroutine")
+			return
+		default:
+		}
+		select {
+		case <-stopCh:
+			klog.Info("Shutting down Hairpin Controller goroutine")
+			return
+		case endpointIP := <-hpc.epC:
+			klog.V(1).Infof("Received request for hairpin setup of endpoint %s, processing", endpointIP)
+			err := hpc.ensureHairpinEnabledForPodInterface(endpointIP)
+			if err != nil {
+				klog.Errorf("unable to set hairpin mode for endpoint %s, its possible that hairpinning will not "+
+					"work as expected. Error was: %v",
+					endpointIP, err)
+			}
+		case <-t.C:
+			healthcheck.SendHeartBeat(healthChan, "HPC")
+		}
+	}
+}
+
+func (hpc *hairpinController) ensureHairpinEnabledForPodInterface(endpointIP string) error {
+	klog.V(2).Infof("Attempting to enable hairpin mode for endpoint IP %s", endpointIP)
+	crRuntime, containerID, err := hpc.nsc.findContainerRuntimeReferences(endpointIP)
+	if err != nil {
+		return err
+	}
+	klog.V(2).Infof("Detected runtime %s and container ID %s for endpoint IP %s", crRuntime, containerID, endpointIP)
+
+	runtime.LockOSThread()
+	defer runtime.UnlockOSThread()
+
+	hostNetworkNSHandle, err := netns.Get()
+	if err != nil {
+		return fmt.Errorf("failed to get namespace due to %v", err)
+	}
+	defer utils.CloseCloserDisregardError(&hostNetworkNSHandle)
+
+	var pid int
+	if crRuntime == "docker" {
+		// WARN: This method is deprecated and will be removed once docker-shim is removed from kubelet.
+		pid, err = hpc.nsc.ln.getContainerPidWithDocker(containerID)
+		if err != nil {
+			return fmt.Errorf("failed to prepare endpoint %s to do direct server return due to %v",
+				endpointIP, err)
+		}
+	} else {
+		// We expect CRI compliant runtimes here
+		// ugly workaround, refactoring of pkg/Proxy is required
+		pid, err = hpc.nsc.ln.getContainerPidWithCRI(hpc.nsc.dsr.runtimeEndpoint, containerID)
+		if err != nil {
+			return fmt.Errorf("failed to prepare endpoint %s to do DSR due to: %v", endpointIP, err)
```

Comment:
you should use `fmt.Errorf("...: %w", ..., err)` instead of %v for error type propagation that you can later check with errors.Is or convert with errors.As